File: cookbook/1593. 拆分字符串使唯一子字符串的数目最大.go
package main

func main() {}

func maxUniqueSplit(s string) int {
	set := map[string]bool{}
	max := 0
	var dfs func(string)
	dfs = func(s string) {
		if len(set)+len(s) < max {
			return
		}
		if len(s) == 0 {
			max = Max(max, len(set))
			return
		}
		for i := 0; i < len(s); i++ {
			curStr := s[:i+1]
			if _, ok := set[curStr]; !ok {
				set[curStr] = true
				dfs(s[i+1:])
				delete(set, curStr)
			}
		}
	}
	dfs(s)
	return max
}

func Max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

/* func maxUniqueSplit(s string) int {
	set := map[string]bool{}
	max := 1
	n := len(s)
	var dfs func(int)
	dfs = func(start int) {
		if len(set)+n < max {
			return
		}
		if start == n {
			max = Max(max, len(set))
			return
		}
		temp := ""
		for i := start; i < n; i++ {
			temp += string(s[i])
			if _, ok := set[temp]; !ok {
				set[temp] = true
				dfs(i + 1)
				delete(set, temp)
			}
		}
	}
	dfs(0)
	return max
}

func Max(x, y int) int {
	if x > y {
		return x
	}
	return y
}
*/

File: leetcode_problems/1593. 拆分字符串使唯一子字符串的数目最大.ts
var maxUniqueSplit = function (s: string): number {
	const S = new Set()
	const n = s.length
	let max = 1
	const dfs = (start: number) => {
		if (n - start + S.size <= max) return
		if (start === n) {
			max = Math.max(max, S.size)
			return
		}
		let temp = ''
		for (let i = start; i < n; i++) {
			temp += s[i]
			if (!S.has(temp)) {
				S.add(temp)
				dfs(i + 1)
				S.delete(temp)
			}
		}
	}
	dfs(0)
	return max
}

{
	const maxUniqueSplit = function (s: string): number {
		const map = new Map<string, boolean>()
		let max = 0
		const dfs = (s: string) => {
			if (map.size + s.length < max) return
			if (s.length === 0) {
				max = Math.max(map.size, max)
			}
			for (let i = 0; i < s.length; i++) {
				const temp = s.substring(0, i + 1)
				if (!map.has(temp)) {
					map.set(temp, true)
					dfs(s.substring(i + 1))
					map.delete(temp)
				}
			}
		}
		dfs(s)
		return max
	}
}
